base/reinterpretarray.jl

```diff
@@ -748,6 +752,90 @@ end
     return true
 end
 
+@assume_effects :foldable function struct_subpadding(::Type{Out}, ::Type{In}) where {Out, In}
+    padding(Out) == padding(In)
+end
+
+@assume_effects :foldable function packedsize(::Type{T}) where T
+    pads = padding(T)
+    return sizeof(T) - sum((p.size for p ∈ pads), init = 0)
+end
+
+@assume_effects :foldable ispacked(::Type{T}) where T = isempty(padding(T))
+
+function _copytopacked!(ptr_out::Ptr{Out}, ptr_in::Ptr{In}) where {Out, In}
+    writeoffset = 0
+    for i ∈ 1:fieldcount(In)
+        readoffset = fieldoffset(In, i)
+        fT = fieldtype(In, i)
+        if ispacked(fT)
+            readsize = sizeof(fT)
+            memcpy(ptr_out + writeoffset, ptr_in + readoffset, readsize)
+            writeoffset += readsize
+        else # nested padded type
+            _copytopacked!(ptr_out + writeoffset, Ptr{fT}(ptr_in + readoffset))
+            writeoffset += packedsize(fT)
+        end
+    end
+end
+
+function _copyfrompacked!(ptr_out::Ptr{Out}, ptr_in::Ptr{In}) where {Out, In}
+    readoffset = 0
+    for i ∈ 1:fieldcount(Out)
+        writeoffset = fieldoffset(Out, i)
+        fT = fieldtype(Out, i)
+        if ispacked(fT)
+            writesize = sizeof(fT)
+            memcpy(ptr_out + writeoffset, ptr_in + readoffset, writesize)
+            readoffset += writesize
+        else # nested padded type
+            _copyfrompacked!(Ptr{fT}(ptr_out + writeoffset), ptr_in + readoffset)
+            readoffset += packedsize(fT)
+        end
+    end
+end
+
+@inline function reinterpret(::Type{Out}, x::In) where {Out, In}
+    isbitstype(Out) || error("reinterpret target type must be isbits")
+    isbitstype(In) || error("reinterpret source type must be isbits")
+    isprimitivetype(Out) && isprimitivetype(In) && return bitcast(Out, x)
```

Comment:
That only works for same sized types, no?

Reply:
Yeah.

We could allow `UInt64 => UInt8` to implicitly mean `UInt64 => NTuple{8, UInt8}`? For any output type that can fit a whole number of times into the input type.

What would `UInt8 => UInt64` do? Still error, or pad with zeros?

Reply:
In FieldFlags.jl I zeroextend when outsize > insize, and truncate when outsize < insize because that's what makes sense in context, but I don't know what Base wants to do.

The first case is a bit odd, because `UInt8` is not a collection, so returning one feels wrong - the existing interface already requires giving an array..

I'd personally err on the side of erroring in the latter case, because reinterpreting that doesn't really make sense if you need to add bits that aren't there. `reinterpret` is not `convert` after all. 

Maybe erroring in both cases is best? It's not like you can reinterpret an `UInt64` to a single `UInt8` :thinking:

Reply:
Yeah, to me it feels like this interface would be best to require the user to be specific and make sure the types match, and make sure the round-trip works.